src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java

```diff
@@ -1161,6 +1153,33 @@ private void validateStaticCodeAnalysisSettings(ProgrammingExercise programmingE
         return ResponseEntity.ok().build();
     }
 
+    /**
+     * PUT /programming-exercises/re-evaluate : Re-evaluates and updates an existing ProgrammingExercise.
+     *
+     * @param programmingExercise the ProgrammingExercise to re-evaluate and update
+     * @param deleteFeedbacks  about checking if the feedbacks should be deleted when the associated grading instructions are deleted
+     *
+     * @return the ResponseEntity with status 200 (OK) and with body the updated ProgrammingExercise, or
+     * with status 400 (Bad Request) if the ProgrammingExercise is not valid, or with status 500 (Internal
+     * Server Error) if the ProgrammingExercise couldn't be updated
+     */
+    @PutMapping(Endpoints.REEVALUATE_EXERCISE)
+    @PreAuthorize("hasRole('EDITOR')")
+    @FeatureToggle(Feature.PROGRAMMING_EXERCISES)
+    public ResponseEntity<ProgrammingExercise> reEvaluateAndUpdateProgrammingExercise(@RequestBody ProgrammingExercise programmingExercise,
+            @RequestParam(value = "deleteFeedbacks", required = false) Boolean deleteFeedbacks) {
+        log.debug("REST request to re-evaluate ProgrammingExercise : {}", programmingExercise);
+
+        // fetch course from database to make sure client didn't change groups
+        var user = userRepository.getUserWithGroupsAndAuthorities();
+        Course course = courseService.retrieveCourseOverExerciseGroupOrCourseId(programmingExercise);
+        authCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.EDITOR, course, user);
```

Comment:
I would prefer this to be consistent with the procedure in `src/main/java/de/tum/in/www1/artemis/web/rest/TextExerciseResource.java`, i.e.:

```
// Check that the user is authorized to update the exercise 
User user = userRepository.getUserWithGroupsAndAuthorities();
if (!authCheckService.isAtLeastEditorInCourse(course, user)) { 
     return forbidden();
}
```

Reply:
hmmm, I did like that at first, but after looking at the other endpoints in the `ProgrammingExerciseResource`, I thought that it would be better if I put this check in here. What do you think, should I change it?

Reply:
makes sense to align it with the other endpoints in this resource. In this case, you can use this statement (as in `unlockAllRepositories` and `lockAllRepositories`) instead of lines 1173-1176 I guess:
```
authCheckService.checkHasAtLeastRoleForExerciseElseThrow(Role.INSTRUCTOR, programmingExercise, null);
```

Reply:
We/I need to do alot of refactoring in the RestControllers, so i would look at the RestController-documentation to see how it should be done.

Apart from that, the current version is shorter and easier to read imo. I would keep it as is.

Reply:
Yes, please use `checkHasAtLeastRoleInCourseElseThrow` instead of manually throwing `forbidden` (we want to avoid that pattern and use the more concise helper method instead).